extractor/filesystem/ffa/unknownbinary/unknownbinary.go

```diff
@@ -0,0 +1,125 @@
+// Package unknownbinary identifies binary files on the filesystem and adds them as packages.
+package unknownbinary
+
+import (
+	"context"
+	"regexp"
+
+	//nolint:gosec //md5 used to identify files, not for security purposes
+	"strings"
+
+	"github.com/google/osv-scalibr/extractor"
+	"github.com/google/osv-scalibr/extractor/filesystem"
+	"github.com/google/osv-scalibr/inventory"
+	"github.com/google/osv-scalibr/plugin"
+	"github.com/google/osv-scalibr/purl"
+)
+
+var (
+	// Common binary / executable file extensions
+	fileExts = []string{
+		".a",
+		// Binary extensions
+		".bin",
+		".elf",
+		".run",
+		".o",
+
+		// Shared library extension
+		".so",
+		// and .so.[number]
+
+		// Script extensions
+		".py", // Python
+		".sh", // bash/sh/zsh
+		".bash",
+
+		".pl",  // Perl
+		".rb",  // Ruby
+		".php", // Php
+		".awk", // Awk
+		".tcl", // tcl
+	}
+	fileExtRegexes = []*regexp.Regexp{
+		regexp.MustCompile(`.so.\d+$`),
+	}
+)
+
+const (
+	// Name is the unique name of this extractor.
+	Name = "filesystem/unknownbinaries"
+)
+
+// Extractor finds unknown binaries on the filesystem
+type Extractor struct {
+}
+
+// Name of the extractor.
+func (e *Extractor) Name() string { return Name }
+
+// Version of the extractor.
+func (e *Extractor) Version() int { return 0 }
+
+// Requirements of the extractor.
+func (e *Extractor) Requirements() *plugin.Capabilities {
+	return &plugin.Capabilities{
+		OS: plugin.OSUnix,
+	}
+}
+
+// FileRequired returns true for likely directories to contain vendored c/c++ code
+func (e *Extractor) FileRequired(fapi filesystem.FileAPI) bool {
+	path := fapi.Path()
+
+	for _, ext := range fileExts {
+		if strings.HasSuffix(path, ext) {
+			return true
+		}
+	}
+
+	for _, regex := range fileExtRegexes {
+		if regex.MatchString(path) {
+			return true
+		}
+	}
+
+	stat, err := fapi.Stat()
+	if err != nil {
+		return false
+	}
+
+	if !stat.Mode().IsRegular() {
+		return false
+	}
+	// Is executable bit set
+	if stat.Mode().Perm()&0o111 != 0 {
```

Comment:
We might want to merge the logic here with filesystem.IsInterestingExecutable - https://github.com/google/osv-scalibr/blob/main/extractor/filesystem/filesystem.go#L684

It uses a denylist instead of an allowlist so maybe we could have an allowlist and a denylist mode for it and move the allowlist from here to filesystem.go

Reply:
Ah neat, did not realize that function existed. I'll have a look at merging that.

Reply:
I am a bit confused by the logic here: https://github.com/google/osv-scalibr/blob/main/extractor/filesystem/filesystem.go#L692

which is making basically all files in windows return true InterestingExecutable

There's also a test here that reinforces that behavior: https://github.com/google/osv-scalibr/blob/50e3de48d73fc20cb387326200b18336d2f2a2f1/extractor/filesystem/filesystem_test.go#L1070

is this intentional?